### PR TITLE
3856: working on support for schema folders

### PR DESCRIPTION
**Pull request by MHove89**
We need to discuss the changes. 

> The import command does not currently support creating the schemas in the correct folder when new schemas are present. This is because we are not setting a name when creating the schema through the import functionality. 
> Should we reconsider how this works? Or even set the name manually in this scenario based on the path? 

@jesperweber  and @MHove89  agreed to generate names based on the path of the schema in the file system. So if schema is placed in the folder "Pages", then the name of the schema in Enterspeed would be Pages/alias

> We have introduced folders for normal schemas in the CLI, and would like to do the same for partials.

Today the logic is that all schemas inside a partial folder is a partial schema. This worked fine before, but since folders  was introduced, this is not ideal anymore. We want to keep it aligned with how it works in the web editor. However we have no way of defining the type of the schema on disk. In the web editor we have a flow for creating the schema in the correct type, as well as meta data about the schema when editing. 

How could we do the same or similar for the CLI? 

Currently schemas are created in a folder structure based on the name. We could do the same for Partials, but still need some way of defining a partial schema, so we can communicate this to the developer, as well as sending the correct information when creating the schema through the management API. 

We have discussed the following

 
1. Adding meta data to the top of the file as comments. However this is not valid JSON.
```
// Schema type: Full
// Schema format: JSON
// Alias: products
// Status: Editable
// Version 11 

{
	"properties": {
		"name": "{item.name}",
		"areas": {
			"type": "array",
			"input": "{item.areas}",
			"items": {
				"type": "partial",
				"input": "{item}",
				"alias": "area"
			}
		}
	}
}
``` 
2. Keeping all partial schemas in the folder Partials, but keeping the same folder structure, but with the root folder as the Partial folder, instead of schemas. This could be misleading, since this would not be how it is presented in the web editor. 

Example
web (Pages/{partialName}
file (Partials/Pages/{partialName}

### CODE REVIEW

**Comment by jesperweber** [comment]
An option 3 could also be to have partials postfixed in the file name, e.g. [schemaAlias]_partial.js

**Comment by jesperweber** [comment]
> I see a couple of options:
> 
> 4. Stop supporting JSON Schemas in new versions of the CLI - because then we can use the type to identify schema type:
> 
> `/** @type {Enterspeed.FullSchema} */`
> 
> 5. Create a .es-state file that has a index of all schemas and their type
> 
> What do you guys think?

Option 4 could work if we are willing to drop the json schema support. 
However it's not just as simple as checking for the type. As I recall the type is not required today it's optional and only there to provide IntelliSense, but maybe it should be required?
Also we have other types of schemas and more will come in the future, meaning that we will have to update the CLI every time.

/** @type {Enterspeed.Destinations.Relewise.Schemas.Product} */
/** @type {Enterspeed.Destinations.Relewise.Schemas.Content} */
/** @type {Enterspeed.Destinations.Relewise.Schemas.Brand} */
/** @type {Enterspeed.Destinations.Relewise.Schemas.ProductCategory} */
/** @type {Enterspeed.Destinations.Relewise.Schemas.ContentCategory} */

Maybe it's primarely the full schemas where we have different types, so maybe we can check for partial and collection type and everything else is full schemas? It's something that we need to consider.

Option 5 could also work. I think it would work really fine when using the CLI commands, but I don't think it's that nice when users start to work with the schemas locally as the schemas and the state file need to be in sync.
What happens if the user deletes or adds a schema file. Then the user also manually need to update the state file?

**Comment by emilras** [comment]
> Option 4 could work if we are willing to drop the json schema support.

> Maybe it's primarely the full schemas where we have different types, so maybe we can check for partial and collection type and everything else is full schemas? It's something that we need to consider.

Yeah. This is not my favourite option, but it could work and deducing the type based on singling out partials and types could work... Its a simple solution but also potentially error prone if we use too many assumptions.
 
> Option 5 could also work. I think it would work really fine when using the CLI commands, but I don't think it's that nice when users start to work with the schemas locally as the schemas and the state file need to be in sync. What happens if the user deletes or adds a schema file. Then the user also manually need to update the state file?

As I recall you need to save the schema using the CLI before you can deploy the schema, so I would think that the CLI needs to be invoked before we need the type of schema...? So I would assume that the CLI have a chance to update the state file before it needs the information.

**Comment by jesperweber** [comment]
> > Option 4 could work if we are willing to drop the json schema support.
> 
> > Maybe it's primarely the full schemas where we have different types, so maybe we can check for partial and collection type and everything else is full schemas? It's something that we need to consider.
> 
> Yeah. This is not my favourite option, but it could work and deducing the type based on singling out partials and types could work... Its a simple solution but also potentially error prone if we use too many assumptions.
> 
> > Option 5 could also work. I think it would work really fine when using the CLI commands, but I don't think it's that nice when users start to work with the schemas locally as the schemas and the state file need to be in sync. What happens if the user deletes or adds a schema file. Then the user also manually need to update the state file?
> 
> As I recall you need to save the schema using the CLI before you can deploy the schema, so I would think that the CLI needs to be invoked before we need the type of schema...? So I would assume that the CLI have a chance to update the state file before it needs the information.

Yes I think that's right, however the import command it will just read all the schemas on disk, so maybe this should give an error  if you have a schema file without corresponding data in the meta file?

**Comment by emilras** [comment]
> Yes I think that's right, however the import command it will just read all the schemas on disk, so maybe this should give an error if you have a schema file without corresponding data in the meta file?

Good. But yes, you are most likely correct in the import scenario. But perhaps @MHove89 can look into the practicalities?

**Comment by MHove89** [comment]
# Folders
Creating names should happen automatically when creating new schemas under import, but what should happen when we update a path locally for a schema when updating under an import?

## Potential solutions
* Warn user that the path does not match the path in Enterspeed and prompt for either updating file path locally, or update path/name in Enterspeed?

@jesperweber

**Comment by emilras** [comment]
> # Folders
> Creating names should happen automatically when creating new schemas under import, but what should happen when we update a path locally for a schema when updating under an import?

Mmmhhh, interesting line of thought. In the Enterspeed app the folders are simply part of the Schema name... so updating the folders on the disk is a rename. And renaming schemas are relatively low risk. I think we should go with the "Update the name, without any prompts in Enterspeed, so the path matches the local path" option.

**Comment by jesperweber** [comment]
> > # Folders
> > Creating names should happen automatically when creating new schemas under import, but what should happen when we update a path locally for a schema when updating under an import?
> 
> Mmmhhh, interesting line of thought. In the Enterspeed app the folders are simply part of the Schema name... so updating the folders on the disk is a rename. And renaming schemas are relatively low risk. I think we should go with the "Update the name, without any prompts in Enterspeed, so the path matches the local path" option.

Agree, updating the name

**Comment by jesperweber** [comment]
We have decided to go with option 3: [schemaAlias].partial.js

If we later need name it can still be included in the file as a comment or in a .es-state file.

**Comment by MHove89** [comment]
Hi @jesperweber 
Do you want to discuss the changes done here, before the review, or would you like to go ahead and test it yourself first?

**Comment by jesperweber** [comment]
After moving a schema from root into a folder and saving the file, the filename gets a backslash and not a forward slash as the app expects.
image
image

**Comment by jesperweber** [comment]
> Hi @jesperweber Do you want to discuss the changes done here, before the review, or would you like to go ahead and test it yourself first?

Just tested some parts of the feature, but yes we could go through it together tomorrow?

File: src/Enterspeed.Cli/Api/MappingSchema/UpdateMappingSchemaRequest.cs [hunk 1]
```diff
@@ -28,10 +26,11 @@ public UpdateMappingSchemaRequestHandler(IEnterspeedClient enterspeedClient)
             _enterspeedClient = enterspeedClient;
         }
 
-        public async Task<UpdateMappingSchemaResponse> Handle(UpdateMappingSchemaRequest updateMappingSchemaRequest, CancellationToken cancellationToken)
+        public async Task<UpdateMappingSchemaResponse> Handle(UpdateMappingSchemaRequest updateMappingSchemaRequest,
+            CancellationToken cancellationToken)
         {
             var request = new RestRequest(
-                $"tenant/mapping-schemas/{updateMappingSchemaRequest.MappingSchemaId}/version/{updateMappingSchemaRequest.Version}",
+                $"tenant/mapping-schemas/{updateMappingSchemaRequest.MappingSchemaId}",
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Api/MappingSchema/UpdateMappingSchemaRequest.cs:33]: I see the endpoint is changed, what are the benefits of this endpoint?

File: src/Enterspeed.Cli/Commands/Schema/ImportSchemaCommand.cs [hunk 2]
```diff
@@ -96,27 +96,57 @@ private async Task<bool> UpdateExistingSchema(SchemaFile schemaFile, QueryMappin
...
         {
+            string name;
+            if (!string.IsNullOrEmpty(schemaFile.RelativeSchemaDirectory))
+            {
+                name = Path.Combine(schemaFile.RelativeSchemaDirectory, schemaFile.Alias);
+            }
+            else
+            {
+                name = schemaFile.Alias;
+            }
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Commands/Schema/ImportSchemaCommand.cs:145]: Should this be handled in the SchemaExtensions class or maybe moved to a SchemaNameService?

File: src/Enterspeed.Cli/Extensions/SchemaExtensions.cs [hunk 3]
```diff
@@ -14,5 +14,18 @@ public static object GetSchemaContent(this SchemaFile schemaFile)
...
+            return schemaName.Contains('/');
+        }
+
+        public static string BuildNewSchemaName(string existingSchemaName, string relativeDirectoryPathOndisk)
+        {
+            var lastSegment = Path.GetFileName(existingSchemaName.TrimEnd(Path.DirectorySeparatorChar));
+            var name = relativeDirectoryPathOndisk + Path.DirectorySeparatorChar + lastSegment;
+
+            return name;
+        }
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Extensions/SchemaExtensions.cs:29]: Maybe move to a SchemaNameService calss?

File: src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs [hunk 4]
```diff
@@ -111,19 +107,43 @@ public SchemaFile GetSchema(string alias, string filePath = null)
...
         }
 
-        return new SchemaFile(alias, schemaType, content, schemaFormat);
+        // Get relative path for schema
+        var currentSchemaDirectoryPath = Path.GetDirectoryName(currentSchemaFilePath);
+        var relativeSchemaDirectoryPath = GetRelativeSchemaPath(currentSchemaDirectoryPath);
+        if (!string.IsNullOrEmpty(relativeSchemaDirectoryPath) && relativeSchemaDirectoryPath.StartsWith(Path.DirectorySeparatorChar))
+        {
+            relativeSchemaDirectoryPath = relativeSchemaDirectoryPath.Remove(0, 1);
+        }
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs:116]: Maybe moved to a SchemaNameService?

File: src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs [hunk 5]
```diff
@@ -137,24 +157,39 @@ public bool SchemaValid(MappingSchemaVersion externalSchema, string schemaAlias)
...
+        if (schemaName != null)
         {
-            Directory.CreateDirectory(PartialSchemaDirectory);
+            if (SchemaExtensions.SchemaIsInDirectory(schemaName))
+            {
+                var schemasRootPath = GetRootSchemaPath();
+                var currentSchemaPathRelative = GetRelativeSchemaPathByName(schemaName);
+                if (!string.IsNullOrEmpty(currentSchemaPathRelative))
+                {
+                    var schemaDirectoryPath = Path.Combine(schemasRootPath, currentSchemaPathRelative);
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs:186]: Maybe move to SchemaNameService?

File: src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs [hunk 6]
```diff
@@ -173,48 +208,47 @@ private static string GetSchemaContent(string alias, string filePath = null)
...
-        return $"{alias}.json";
+        return Path.Combine(SchemaDirectory, GetFileName(alias, format, schemaType));
+    }
+
+    private static string GetFileName(string alias, string format, SchemaType schemaType)
+    {
+        var schemaTypeName = schemaType.ToString().ToLowerInvariant();
+        var schemaName = schemaType.Equals(SchemaType.Partial) ? $"{alias}.{schemaTypeName}" : $"{alias}";
+        return format.Equals(SchemaConstants.JavascriptFormat) ? $"{schemaName}.js" : $"{schemaName}.json";
     }
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs:234]: Maybe moved to SchemaNameService?

**Comment by MHove89** [comment]
> After moving a schema from root into a folder and saving the file, the filename gets a backslash and not a forward slash as the app expects. image image

I am not able to replicate this. Tested in Linux and Windows. Could you show me how you did this?

**Comment by MHove89** [comment]
@jesperweber I have worked on cleaning up the schemaFileService, as well as some naming. Bug with "/" in front of filename has been fixed. .full added to files for schemas of type full

File: src/Enterspeed.Cli/Services/FileService/FilePathService.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        return relativeSchemaDirectoryPath;
+    }
+
+    public string GetAliasFromFilePath(string filePath)
+    {
+        return Path.GetFileNameWithoutExtension(filePath).Replace(".partial", "")
+            .Replace(".full", "");
+    }
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/FilePathService.cs:47]: Should this be moved to the SchemaName Service, together with the `GetFileName` from the SchemaFileService? Then we don't have logic two places in regards to the schema type in the name?

File: src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs [hunk 8]
```diff
@@ -202,53 +180,49 @@ private static void DeleteSchema(string alias)
...
+            var schemaDirectoryPath = _filePathService.GetDirectoryPathBySchemaName(schemaName);
+            var fullFilePath = Path.Combine(schemaDirectoryPath, GetFileName(alias, format, schemaType));
+            return fullFilePath;
         }
 
-        return Path.Combine(SchemaDirectory, GetFileName(alias, format, schemaType));
+        return Path.Combine(_filePathService.GetRelativeRootDirectoryPath(), GetFileName(alias, format, schemaType));
     }
 
     private static string GetFileName(string alias, string format, SchemaType schemaType)
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/SchemaFileService.cs:202]: Should this be moved to the SchemaNameService?

File: src/Enterspeed.Cli/Services/FileService/FilePathService.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+        if (!string.IsNullOrEmpty(relativeSchemaDirectoryPath) && relativeSchemaDirectoryPath.StartsWith(Path.DirectorySeparatorChar))
+        {
+            relativeSchemaDirectoryPath = relativeSchemaDirectoryPath.Remove(0, 1);
+        }
+
+        return relativeSchemaDirectoryPath;
+    }
+
+    public string GetAliasFromFilePath(string filePath)
```
Comment (jesperweber) [inline src/Enterspeed.Cli/Services/FileService/FilePathService.cs:43]: Should this be moved to the SchemaName Service, together with the GetFileName from the SchemaFileService? Then we don't have logic two places in regards to the schema type in the name?

**Review by jesperweber — COMMENTED** [review]
Code is way more clean now 💪

I still have issues with backslash instead of forward slash.

cliFolder

**Pull request merged**